searchcore/src/vespa/searchcore/proton/server/documentretriever.h

```diff
@@ -25,15 +25,26 @@ class DocumentRetriever : public DocumentRetrieverBase {
                       const search::IDocumentStore &doc_store);
     ~DocumentRetriever() override;
 
-    document::Document::UP getDocument(search::DocumentIdT lid) const override;
+    document::Document::UP getDocumentByLidOnly(search::DocumentIdT lid) const override;
     void visitDocuments(const LidVector & lids, search::IDocumentVisitor & visitor, ReadConsistency) const override;
+    DocumentUP getDocument(search::DocumentIdT lid, const document::DocumentId &, const document::FieldSet &) const override;
     void populate(search::DocumentIdT lid, document::Document & doc) const;
+    bool needFetchFromDocStore(const document::FieldSet &) const;
 private:
+    using FieldSetAttributeMap = vespalib::hash_map<uint64_t, bool>;
+    bool needFetchFromDocStore(uint64_t key, const document::Field &) const;
+    bool needFetchFromDocStore(uint64_t key, const document::Field::Set &) const;
+    void populate(search::DocumentIdT lid, document::Document & doc, const document::Field::Set & attributeFields) const;
+
+    bool isFieldAttribute(const document::Field & field) const;
     const search::index::Schema     &_schema;
     const search::IAttributeManager &_attr_manager;
     const search::IDocumentStore    &_doc_store;
     PositionFields                   _possiblePositionFields;
-    AttributeFields                  _attributeFields;
+    document::Field::Set             _attributeFields;
+    bool                             _areAllFieldsAttributes;
+    mutable FieldSetAttributeMap     _isFieldSetAttributeOnly;
```

Comment:
if this is actually useful/needed, please refactor into a separate class telling us if a specific field set is a superset of other field sets or not. How much faster is it to take a lock compared to looking up a set of strings in an immutable hash table?

Reply:
Fixed